File: 백준/Gold/20058. 마법사 상어와 파이어스톰/마법사 상어와 파이어스톰.java
import java.util.*;
import java.io.*;
import java.math.*;

public class Main {

    static BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

    static int N, Q;

    static int[][] board;
    static int[] levels;

    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};
    static boolean[][] visited;
    static int[][] ice_count;
    static int total_ice;

    static int linked_ice;
    static int max_ice;

    static void init() throws Exception {
        StringTokenizer tkn = new StringTokenizer(reader.readLine());
        N = (int) Math.pow(2, Integer.parseInt(tkn.nextToken())); // 보드 크기. 2 ~ 6 -> 4 ~ 64
        Q = Integer.parseInt(tkn.nextToken()); // 파이어스톰 시전 횟수. 1 ~ 1천

        // 얼음판 입력받기
        board = new int[N][N];
        for(int i=0; i<N; i++) {
            tkn = new StringTokenizer(reader.readLine());
            for(int j=0; j<N; j++) {
                board[i][j] = Integer.parseInt(tkn.nextToken()); // 얼음 크기. 1 ~ 100
            }
        }

        // 시전 레벨 입력받기
        levels = new int[Q];
        tkn = new StringTokenizer(reader.readLine());
        for(int i=0; i<Q; i++) {
            // 시전 레벨. 1 ~ 2^(N)
            levels[i] = (int) Math.pow(2, Integer.parseInt(tkn.nextToken()));
        }

        visited = new boolean[N][N];
        ice_count = new int[N][N];
    }

    // 시작 row, 시작 col부터 크기 siz 만큼의 board 영역을 시계방향으로 90도 회전시킨다.
    static void tornado(int start_row, int start_col, int siz) {

        int[][] tmpBoard = new int[siz][siz];
        for(int r=0; r<siz; r++) {
            for(int c=0; c<siz; c++) {
                tmpBoard[r][c] = board[start_row+r][start_col+c];
            }
        }

        // 1. 행, 열 바꾸기
        for(int r=0; r<siz; r++) {
            for(int c=r+1; c<siz; c++) {
                int tmp = tmpBoard[r][c];
                tmpBoard[r][c] = tmpBoard[c][r];
                tmpBoard[c][r] = tmp;
            }
        }

        // 2. 좌우 뒤집기
        for(int r=0; r<siz; r++) {
            for(int c=0; c<siz/2; c++) {
                int tmp = tmpBoard[r][c];
                int opposite_col = siz-c-1;
                tmpBoard[r][c] = tmpBoard[r][opposite_col];
                tmpBoard[r][opposite_col] = tmp;
            }
        }

        for(int r=0; r<siz; r++) {
            for(int c=0; c<siz; c++) {
                board[start_row+r][start_col+c] = tmpBoard[r][c];
            }
        }

    }

    // 얼음 녹일 때 쓰는 dfs
    static void dfs_fire(int row, int col) {
        visited[row][col] = true;

        int nr, nc;
        for(int i=0; i<4; i++) {
            nr = row + dr[i];
            nc = col + dc[i];
            if (nr >=0 && nc >=0 && nr < N && nc <N) {
                if (board[nr][nc] > 0) ice_count[row][col]++; // 얼음이 있으면 카운트
                if (!visited[nr][nc]) dfs_fire(nr, nc);
            }
        }

    }

    // 사방에 둘러싼 얼음이 3개 미만인 경우, 해당 얼음의 크기를 1 감소
    static void fireball() {
        // 방문 배열, 녹는 여부 배열 초기화
        for(int i=0; i<N; i++) {
            Arrays.fill(visited[i], false);
            Arrays.fill(ice_count[i], 0);
        }

        // 녹는 얼음 체크
        dfs_fire(0, 0);

        // 얼음 녹이기
        for(int r=0; r<N; r++) {
            for(int c=0; c<N; c++) {
                if (ice_count[r][c] < 3) board[r][c] = Math.max(0, board[r][c]-1);
            }
        }

    }

    static void fireStorm(int level) {
        // 1. 얼음판을 시전 레벨만큼의 부분 행렬로 나눈 다음, 시계방향으로 90도 회전시킨다.
        for(int i=0; i<N; i+=level) {
             for(int j=0; j<N; j+=level) {
                 tornado(i, j, level);
             }
        }

        // 2. 사방에 둘러싼 얼음이 3개 미만인 경우, 해당 얼음의 크기를 1 감소시킨다.
        fireball();
    }

    static void dfs_calc(int row, int col) {
        linked_ice++;
        visited[row][col] = true;

        int nr, nc;
        for(int i=0; i<4; i++) {
            nr = row + dr[i];
            nc = col + dc[i];
            if (nr >=0 && nc >=0 && nr<N && nc<N && !visited[nr][nc]) {
                if (board[nr][nc] > 0) {
                    dfs_calc(nr, nc);
                }
            }
        }
    }

    static String calc() {
        // 1. 지정한 횟수만큼 fireStorm 실행
        for(int i=0; i<Q; i++) {
            fireStorm(levels[i]);
        }

        // 2. A[r][c]의 전체 합 구하기
        // 3. 남은 얼음 중 가장 큰 얼음 덩이 구하기
        // 방문 배열 초기화
        for(int i=0; i<N; i++) Arrays.fill(visited[i], false);
        max_ice = 0;
        total_ice = 0;

        for(int r=0; r<N; r++) {
            for(int c=0; c<N; c++) {
                total_ice += board[r][c];
                if(!visited[r][c] && board[r][c] > 0) {
                    linked_ice = 0;
                    dfs_calc(r, c);
                    max_ice = Math.max(max_ice, linked_ice);
                }
            }
        }

        if (max_ice < 2) max_ice = 0;

        StringBuilder sb = new StringBuilder();
        sb.append(total_ice).append('\n');
        sb.append(max_ice).append('\n');

        return sb.toString();
    }

    public static void main(String[] args) throws Exception {
        init();
        String result = calc();
        System.out.println(result);
    }

}

File: 백준/Gold/20058. 마법사 상어와 파이어스톰/마법사 상어와 파이어스톰.cc
#include <iostream>
#include <cmath>
#include <vector>

using namespace std;

int N, Q;
vector<vector<int>> board;
vector<int> levels;

int dr[] = {-1, 1, 0, 0};
int dc[] = {0, 0, -1, 1};
vector<vector<bool>> visited;
vector<vector<bool>> is_melt;
int max_ice;
int total_ice;
int cur_ice;

void init() {
    cin >> N >> Q;
    N = (int)pow(2, N); // 줄 수. 1 ~ 6 -> (2 ~ 64)

    // new int[N][N]
    // 얼음판 입력받기
    board.resize(N);
    for(int i=0; i<N; i++) {
        board[i].resize(N);
        for(int j=0; j<N; j++) {
            cin >> board[i][j];
        }
    }
    
    // 시전 레벨 입력받기
    levels.resize(Q);
    int tmp;
    for(int i=0; i<Q; i++) {
        cin >> tmp; // 시전 레벨은 1 ~ N. -> (2 ~ 2^N)
        levels[i] = pow(2, tmp);
    }

    // dfs에 쓸 visited, is_melt 크기 초기화
    visited.resize(N);
    is_melt.resize(N);
    for(int i=0; i<N; i++) {
        visited[i].resize(N);
        is_melt[i].resize(N);
    }
}

void tornado(int sr, int sc, int siz) {
    int tmp_board[siz][siz];
    // 입력받기
    for(int r=0; r<siz; r++) {
        for(int c=0; c<siz; c++) {
            tmp_board[r][c] = board[sr+r][sc+c];
        }
    }

    // 1. 행, 열 뒤집기
    for(int r=0; r<siz; r++) {
        for(int c=r+1; c<siz; c++) {
            // r, c swap
            int tmp = tmp_board[r][c];
            tmp_board[r][c] = tmp_board[c][r];
            tmp_board[c][r] = tmp;
        }
    }

    // 2. 좌우 뒤집기
    for(int r=0; r<siz; r++) {
        for(int c=0; c<siz/2; c++) {
            int opp_col = siz-c-1;
            // swap
            int tmp = tmp_board[r][c];
            tmp_board[r][c] = tmp_board[r][opp_col];
            tmp_board[r][opp_col] = tmp;
        }
    }

    // 변경사항 다시 입력하기
    for(int r=0; r<siz; r++) {
        for(int c=0; c<siz; c++) {
            board[sr+r][sc+c] = tmp_board[r][c];
        }
    }
}

// dfs 돌려서 사방에 3군데 이상 얼음이 있는지 확인
void fire_ball(int row, int col) {
    visited[row][col] = true;
    int ice_count = 0;

    for(int i=0; i<4; i++) {
        int nr = row + dr[i];
        int nc = col + dc[i];
        if (nr >=0 && nc >=0 && nr < N && nc < N) {
            if (board[nr][nc] > 0) ice_count++;
            if (!visited[nr][nc]) fire_ball(nr, nc);
        }
    }

    // 주변을 둘러싼 얼음이 3개 미만인 경우 녹는 얼음으로 지정
    if (ice_count < 3) is_melt[row][col] = true;
}

void fire_storm(int level) {
    // 1. 토네이도로 얼음판 회전
    for(int row=0; row<N; row+=level) {
        for(int col=0; col<N; col+=level) {
            tornado(row, col, level);
        }
    }

    // 2. fireball로 조건에 따라 얼음 제거
    for(int i=0; i<N; i++) {
        fill(visited[i].begin(), visited[i].end(), false);
        fill(is_melt[i].begin(), is_melt[i].end(), false);
    }

    fire_ball(0, 0);

    for(int row=0; row<N; row++) {
        for(int col=0; col<N; col++) {
            if (is_melt[row][col]) board[row][col] = max(0, board[row][col]-1);
        }
    }
}

void find_ice_group(int row, int col) {
    visited[row][col] = true;

    if(board[row][col] == 0) return; // 얼음이 없으면 바로 리턴

    total_ice += board[row][col];
    cur_ice++;

    for(int i=0; i<4; i++) {
        int nr = row + dr[i];
        int nc = col + dc[i];
        if (nr >=0 && nc >=0 && nr < N && nc < N && !visited[nr][nc]) {
            find_ice_group(nr, nc);
        }
    }
}

string calc() {
    // 1. Q번 파이어스톰을 시전한다.
    for(int i=0; i<Q; i++) {
        fire_storm(levels[i]);
    }

    // 2. 남은 얼음들의 합을 구한다.
    // 3. 남아있는 얼음 중 가장 큰 덩어리의 크기를 구한다. (dfs)
    for(int i=0; i<N; i++)
        fill(visited[i].begin(), visited[i].end(), false);

    max_ice = 0;
    total_ice = 0;
    for(int i=0; i<N; i++) {
        for(int j=0; j<N; j++) {
            if (!visited[i][j]) {
                cur_ice = 0;
                find_ice_group(i, j);
                max_ice = max(max_ice, cur_ice);
            }
        }
    }

    if (max_ice < 2) max_ice = 0;

    string result;
    result += to_string(total_ice) + "\n";
    result += to_string(max_ice);

    return result;
}

int main() {

    init();

    string result = calc();
    cout << result << endl;

    return 0;
}
